hw/dv/dpi/common/tcp_server/tcp_server.c

```diff
@@ -178,7 +178,10 @@ static int client_tryaccept(struct tcp_server_ctx *ctx) {
   int rv;
 
   assert(ctx->sfd > 0);
-  assert(ctx->cfd == 0);
+
+  // assume a connection has been lost if ctx->cfd > 0
+  if (ctx->cfd > 0)
+    tcp_server_client_close(ctx);
```

Comment:
I've just taken another look at this and I *think* I understand what's going on.

The code in `server_create` selects between a client and a server socket, waiting for them to be ready. If there is data on the server socket, it calls `client_tryaccept`. This function tries to `accept()` the server connection, putting the file descriptor for the connected socket into `ctx->cfd`.

I think the idea is that this is a "one-shot" process: once we've set up the connection, we don't expect anything new on `ctx->sfd`. We then try to communicate through `ctx->cfd` until someone calls `tcp_server_close`.

If new data appears on `ctx->sfd`, things are going to go badly wrong, and we'll hit this assertion. (If you're seeing this locally, would you mind adding a couple of debug prints to check that's what happening?)

What I don't understand is what we *should* do in this situation. I'm hoping that someone who knows more about socket programming than me can chime in! In particular, is this a situation that we expect to happen? Or is the correct fix to change `server_create` so that it does something like this?
```c
  while (ctx->socket_run) {
    // Initialise structure of fds
    fd_set read_fds;
    FD_ZERO(&read_fds);
    if (ctx->sfd && !ctx->cfd) {  // CHANGE HERE!
      FD_SET(ctx->sfd, &read_fds);
    }
    if (ctx->cfd) {
      FD_SET(ctx->cfd, &read_fds);
    }
    ...
```
That way we'd only ever accept the first connection.

The code in question dates back to the "start of time" in our repository, where it was originally designed just to handle a virtual JTAG connection with OpenOCD. It's possible that you're using it in a more general context than it was designed to cope with? (Not a problem, but it might explain why it doesn't work!)

Reply:
Hi, yeah, that kinda matches my understanding although this is my first foray into socket programming.  I'm using it for exactly the same use case, to drive SWD pins from a remote pyOCD connection.  However, I got fed up with the Verilator model crashing when closing the connection and trying to reconnect.  I can understand why it was originally designed to only accept one connection though - you expect the model to be spun up, connect via OpenOCD, run a test and close down.

Reply:
Ah, now I understand. In which case, this seems like reasonable behaviour. Would you mind adding a comment explaining what exactly we're doing and what we'd expect to happen ("if there's a new connection request, boot off anything that we were talking to before and start communicating on the new channel" or something)